File: ecommerce/api/model/livestream.go
package model

import (
	"ecommerce/internal/database/gen/model"
	"time"
)

type GetLivestreamsQueryParam struct {
	Status []string `form:"status[]"`
	ShopId int64    `form:"shop_id,default=0"`
	PaginationParam
}

type LivestreamProductCreateRequest struct {
	IDProduct          int64 `json:"id_product"`
	Priority           int32 `json:"priority"`
	LivestreamVariants []*struct {
		IDVariant                  int64 `json:"id_variant"`
		LivestreamExternalVariants []*struct {
			IDExternalVariant int64 `json:"id_external_variant"`
			Quantity          int32 `json:"quantity"`
		} `json:"livestream_external_variants"`
	} `json:"livestream_variants"`
}

type CreateLivestreamRequest struct {
	Title              string                            `json:"title"`
	Description        string                            `json:"description"`
	StartTime          *time.Time                        `json:"start_time"`
	EndTime            *time.Time                        `json:"end_time"`
	LivestreamProducts []*LivestreamProductCreateRequest `json:"livestream_products"`
}

type UpdateLivestreamRequest struct {
	Title       *string    `json:"title"`
	Description *string    `json:"description"`
	Status      *string    `json:"status"`
	MeetingID   *string    `json:"meeting_id"`
	HlsURL      *string    `json:"hls_url"`
	StartTime   *time.Time `json:"start_time"`
	EndTime     *string    `json:"end_time"`
}

type SetLivestreamHlsRequest struct {
	HlsUrl string `json:"hls_url"`
}

type GetLivestreamResponse struct {
	IDLivestream int64   `json:"id_livestream" xml:"id_livestream"`
	IDShop       int64   `json:"id_shop" xml:"fk_shop"`
	Title        string  `json:"title" xml:"title"`
	Description  *string `json:"description" xml:"description"`
	Status       string  `json:"status" xml:"status"`
	MeetingID    string  `json:"meeting_id" xml:"meeting_id"`
	HlsURL       *string `json:"hls_url" xml:"hls_url"`
	IsHost       bool    `json:"is_host"`
	ShopName     string  `json:"shop_name"`
	IsFollowed   bool    `json:"is_followed"`
}

type UpdateLivestreamExternalVariantQuantityRequest []*struct {
	LivestreamExternalVariantId int64 `json:"id_livestream_external_variant"`
	Quantity                    int32 `json:"quantity"`
}

type UpdateLivestreamProductPriorityRequest []*struct {
	LivestreamProductId int64 `json:"id_livestream_product"`
	Priority            int32 `json:"priority"`
}

type UpdateLivestreamProductsRequest []*struct {
	IDLivestreamProduct int64  `json:"id_livestream_product"`
	Priority            *int32 `json:"priority"`
	IsLivestreamed      *bool  `json:"is_livestreamed"`
}

type RegisterLivestreamProductFollowerRequest struct {
	IDLivestreamProducts []int64 `json:"id_livestream_products"`
	IDUser               int64   `json:"id_user"`
}

type LivestreamProductFollowerDTO struct {
	UserIds    []int64           `json:"user_ids"`
	Livestream *model.Livestream `json:"livestream"`
	Product    *model.Product    `json:"livestream_product"`
}

File: web/src/api/livestream.ts
import axios from "./axios";
import qs from "qs";

import { IBaseLivestream, IBaseLivestreamProduct } from "@/model/livestream";

export const getLivestreams = async (queryParam: { [key: string]: any }) => {
  const res = await axios.get<IBaseLivestream[]>(
    `livestreams/?${decodeURIComponent(qs.stringify(queryParam, { arrayFormat: "brackets" }))}`,
  );
  return res.data;
};

export interface ILivestream {
  id_livestream: number;
  fk_shop: number;
  title: string;
  description: string;
  status: string;
  meeting_id: string;
  hls_url: string;
}

export interface ILivestreamExternalVariant {
  id_external_variant: number;
  quantity: number;
}

export interface ILivestreamProduct {
  id_product: number;
  priority: number;
  livestream_variants: {
    id_variant: number;
    livestream_external_variants: ILivestreamExternalVariant[];
  }[];
}

interface ILivestreamWithProducts extends ILivestream {
  livestream_products: ILivestreamProduct[];
}

export const createLivestream = async (
  shopId: number,
  data: ILivestreamWithProducts,
) => {
  return axios.post(`shops/${shopId}/livestreams/create`, data);
};

export const getLivestreamProducts = async (livestreamId: number) => {
  const res = await axios.get<IBaseLivestreamProduct[]>(
    `livestreams/${livestreamId}/livestream_products`,
  );
  return res.data;
};

interface IGetLivestreamResponse extends IBaseLivestream {
  is_host: boolean;
  shop_name: string;
}

export const getLivestream = async (livestreamId: number) => {
  const res = await axios.get<IGetLivestreamResponse>(
    `livestreams/${livestreamId}`,
  );
  return res.data;
};

interface IUpdateLivestreamRequest {
  title: string;
  description: string;
  status: string;
  meeting_id: string;
  hls_url: string;
  start_time: string;
  end_time: string;
}
export const updateLivestream = async (
  livestreamId: number,
  request: IUpdateLivestreamRequest,
) => {
  const res = await axios.patch(`livestreams/${livestreamId}`, request);
  return res.data;
};

export const saveHls = async (livestreamId: number, hlsUrl: string) => {
  const res = await axios.post(`livestreams/${livestreamId}/save_hls`, {
    hls_url: hlsUrl,
  });
  return res.data;
};

export const addLivestreamProduct = async (
  livestreamId: number,
  data: ILivestreamProduct[],
) => {
  const res = await axios.post(
    `livestreams/${livestreamId}/add_livestream_product`,
    data,
  );
  return res.data;
};
